attestation-agent/attestation-agent/src/eventlog/mod.rs

```diff
@@ -79,124 +93,141 @@ impl EventLog {
         tokio::fs::create_dir_all(EVENTLOG_PARENT_DIR_PATH)
             .await
             .context("create eventlog parent dir")?;
-        if Path::new(EVENTLOG_PATH).exists() {
-            debug!("Previous AAEL found. Skip INIT entry recording...");
-            let content = tokio::fs::read_to_string(EVENTLOG_PATH)
-                .await
-                .context("Read AAEL")?;
-
-            // The content of AAEL can be empty when the previous AA created this file
-            // but did not do anything.
-            if content.is_empty() {
-                let file = File::options()
-                    .write(true)
-                    .open(EVENTLOG_PATH)
-                    .context("open eventlog")?;
-                let mut eventlog = Self {
-                    writer: Box::new(FileWriter { file }),
+        let mut file = File::options()
+            .append(true)
+            .create(true)
+            .open(EVENTLOG_PATH)
+            .context("open AAEL file")?;
+        let pos = file.stream_position()?;
+
+        let mut writer = Box::new(FileWriter { file, pos });
+        // if any WAL cache file exists, we should handle recovering from crash
+        match Self::read_wal_cache(alg.digest_len()) {
+            Ok(Some(wal_cache)) => {
+                warn!("Recover from a previous crash.");
+                let current_pcr = rtmr_extender.get_runtime_measurement(pcr).await.context("get runtime measurement")?;
+                let aael_event = Event::try_from(&wal_cache.event_data[..])?;
+                let (tcg2_event, tcg2_event_digest) = Into::<Tcg2EventEntry>::into(aael_event).digest(alg);
+                let tcg2_event_data = tcg2_event.to_le_bytes();
+
+                // if the PCR has not been extended yet, we should just write eventlog
+                if current_pcr != wal_cache.expected_pcr {
+                    let mut pcr_status = current_pcr.clone();
+                    let mut tcg2_event_digest_clone = tcg2_event_digest.clone();
+                    pcr_status.append(&mut tcg2_event_digest_clone);
+                    let digest_to_be_updated = alg.digest(&pcr_status);
+
+                    if digest_to_be_updated != wal_cache.expected_pcr {
+                        bail!("fatal error when recovering. The eventlog file {EVENTLOG_PATH} is probably corrupted, or other process has extend the target PCR {pcr}.")
+                    }
+
+                    // else, update the PCR
+                    rtmr_extender.extend_runtime_measurement(tcg2_event_digest, pcr).await?;
+                }
```

Comment:
Should we be cleaning up the WAL file somewhere around here since we have just synced up the state?